### PR TITLE
cmd/shfmt: implement --from-json and improve -tojson

### LINKED ISSUE #35: cmd/shfmt: add flags to read/write AST in JSON format to simplify integration with other languages
Quite a few projects out there implement their own shell/bash parsers. Not only the shells themselves, but also some programs that work with shell scripts such as linters.

It's really a little silly for each one of them, especially tooling around shell scripts, to have to implement their own. Sadly, the shells out there don't allow their use as libraries (that I know of).

This is a library, so it does solve that problem. But being written in Go, it's not easily usable from other languages like Python or Shell itself.

Some ideas come to mind:
- Building the packages as shared libraries, linkable by C: http://blog.ralch.com/tutorial/golang-sharing-libraries/
- Adding an executable alongside `shfmt` that parses an input file and outputs the AST in json or some other easy-to-import format

Both of those could be extended to also include the printer.

Any input will be appreciated, especially from people who would use something like this.

### PR DESCRIPTION
**Pull request by mvdan**
(see commit message)

Fixes #35 again, as we never implemented the "read JSON" side.

### CODE REVIEW

**Comment by mvdan** [comment]
cc @cilki @JounQin

**Comment by mvdan** [comment]
A large portion of the diff is JSON output in test files, which was automatically generated via `go test -u`. It shouldn't be closely reviewed, but I did skim it for any mistakes or unexpected changes.

File: cmd/shfmt/main.go [hunk 1]
```diff
@@ -168,7 +169,8 @@ Printer options:
 Utilities:
 
   -f, --find   recursively find all shell files and print the paths
-  --tojson     print syntax tree to stdout as a typed JSON
+  --to-json    print syntax tree to stdout as a typed JSON
```
Comment (riacataquian) [inline cmd/shfmt/main.go:172]: does it make sense to mention here the flag change, eg `available since v3`? perhaps to make the change obvious.

**Review by riacataquian — APPROVED** [review]
neat :) LGTM!

**Comment by JounQin** [comment]
So it won't a core API but left in `shfmt` right? I'd like to use it without `shfmt` for small bundle, should I copy/paste related codes instead?

File: cmd/shfmt/main.go (hunk 1 above)
Comment (mvdan) [reply cmd/shfmt/main.go:172]: We generally record that info via the changelog. I'm not sure whether including that in the help text or man page helps. If you run `shfmt -h` and the flag is there, then you can use it. If the flag isn't there, then your version is too old, so you can't use it.

**Comment by JounQin** [comment]
@mvdan Thanks for clarifying, I just want to make sure it will be in core API in the future.

**Comment by mvdan** [comment]
Did one last push as I forgot to update the man page; see https://github.com/mvdan/sh/compare/46e0c4e09353c070bc3e23b1aac4a3b800c64d4b..e5d80446cfdede8b096dd7c9553edf1e73ce5ab7. I'll be merging on green. Thanks for the review!

**Pull request merged**